Multi-level Nested Sort with Filters

mattweber opened the pull request:
Allow multiple levels of nested sorting where each level can have it's own filter. This allows us to have complex nested sort criteria just like we have with queries and aggregations.   This is backwards compatible with previous single-level nested sort.

@martijnvg @jpountz Can you please review?

elasticmachine commented:
Since this is a community submitted pull request, a Jenkins build has not been kicked off automatically. Can an Elastic organization member please verify the contents of this patch and then kick off a build manually?

martijnvg commented on docs/reference/search/request/sort.asciidoc:
```diff
@@ -115,24 +115,26 @@ POST /_search
-existing sort options:
+field support has a `nested` sort option with the following properties:
 
-`nested_path`::
```
The migration should document that `nested_path` and `nested_filter` are deprecated.

martijnvg commented on core/src/main/java/org/elasticsearch/search/sort/FieldSortBuilder.java:
```diff
@@ -334,6 +364,7 @@ public static FieldSortBuilder fromXContent(XContentParser parser, String fieldN
         PARSER.declareString((b, v) -> b.order(SortOrder.fromString(v)) , ORDER_FIELD);
         PARSER.declareString((b, v) -> b.sortMode(SortMode.fromString(v)), SORT_MODE);
         PARSER.declareObject(FieldSortBuilder::setNestedFilter, (p, c) -> SortBuilder.parseNestedFilter(p), NESTED_FILTER_FIELD);
```
I think we should add some deprecation logging here too for the `nested_filter` and `nested_path` parameters.

```
private static final DeprecationLogger DEPRECATION_LOGGER = new DeprecationLogger(Loggers.getLogger(ParseField.class));
...
PARSER.declareString((fieldSortBuilder, nestedPath) -> {
            DEPRECATION_LOGGER.deprecated("[nested_path] has been deprecated in favor of the [nested] parameter");
            fieldSortBuilder.setNestedPath(nestedPath);
        } , NESTED_PATH_FIELD);
...
PARSER.declareObject(FieldSortBuilder::setNestedFilter, (p, c) -> {
            DEPRECATION_LOGGER.deprecated("[nested_filter] has been deprecated in favour for the [nested] parameter");
            return SortBuilder.parseNestedFilter(p);}, NESTED_FILTER_FIELD);
...
```

martijnvg commented:
test this please

jpountz commented on core/src/main/java/org/elasticsearch/search/sort/SortBuilder.java:
```diff
@@ -69,6 +78,132 @@
         PARSERS = unmodifiableMap(parsers);
     }
 
+    public static class NestedSort implements Writeable, Reader<NestedSort>, ToXContentObject {
```
Even though it is not strictly a sort builder, can you rename it NestedSortBuilder for consistency and move it to its own class?

mattweber commented:
@martijnvg Thanks for the review!  Thats a bummer about waiting for 6.1, I was hoping it could make a 5.x release since it is backward compatible.  I have addressed your comments, please let me know if you would like any other changes.

jpountz commented on core/src/main/java/org/elasticsearch/search/sort/SortBuilder.java:
```diff
@@ -69,6 +78,132 @@
+        private QueryBuilder filter;
+        private NestedSort nestedSort;
+
+        public NestedSort() {
```
should we requite the path in the constructor since it is mandatory?

mattweber commented:
Thanks @jpountz, I addressed your comments.  How does this look?

martijnvg commented:
retest this please

martijnvg commented:
@mattweber Thanks for fixing the comments. The PR looks good! Only the build failed with checkstyle violations. Can you fix this?

mattweber commented:
@martijnvg @jpountz Thank you for reviewing everything.  Just pushed checkstyle and test fixes.  I think this will do it.

martijnvg commented:
retest this please

martijnvg commented:
The mixed cluster test failures that occurred in the build this PR triggered are related to the fact that the changes aren't yet in the 6.x branch. This causes serialization errors when a 7.0.0-alph1 node reads the shard level search request from a 6.1.0-SNAPSHOT node. So I'll merge this change both to master and 6.x branches at the same time to minimize the regular builds from running into this failure.

jasontedor commented:
The versions and skips should be set so that they cause no failures, backported with version updates, then another commit to master moving the version to the 6.x version they are compatible with.

jasontedor commented:
Here's another PR where I [expand](https://github.com/elastic/elasticsearch/pull/26166#issuecomment-322140800) on this strategy; this is build failure minimizing strategy, not merging as quickly as possible. Instead, we should take our time to get a green build on master, then backport (with the version bumps) and getting a green build again, and then a final commit to get the versions right still maintaining green status.

martijnvg commented:
@jasontedor Good point. I'll follow the same steps mentioned in that PR.

martijnvg commented:
I merged squashed the change into the master branch via 6377afa6c376e4c7d0d477e1d4e5f9dbd222a3c4
and set the version to 7.0.0-alph1 via 472a5dd56b4360037dc7ce7c2837c929b3c3fd25

I'll backport this change to 6.x branch later today unless a build failure related to this change occurs.

@mattweber Thanks for this contribution to nested sorting!

martijnvg commented:
I just realized that as part of the git merge (git merge --squash) that I became the author of this change :( 
Not sure what would be best to restore this mistake. (I can only think of reverting the commit and then commit it again with the @mattweber as author)

jasontedor commented:
@martijnvg For future reference, you can push commits to PRs so you could have pushed the version-changing commit to this PR, then merged from the GitHub UI (preserving @mattweber as the author). As far as correcting, I think that you can:
 - revert the commits that you pushed to master
 - reopen this PR
 - push the version-changing commit to this PR
 - merge this PR from the GitHub UI

elasticmachine commented:
Since this is a community submitted pull request, a Jenkins build has not been kicked off automatically. Can an Elastic organization member please verify the contents of this patch and then kick off a build manually?

martijnvg commented:
@mattweber Can you change the 6.1.0 version check to 7.0.0-alpha1? (as it turns out I can't make this change in your pr) After that I'll hit the squash and merge button and do they backporting to 6.x branch.

jpountz commented:
@martijnvg I'm wondering whether that would work with the online editor, eg. https://github.com/mattweber/elasticsearch/edit/multi_nested_sort/core/src/main/java/org/elasticsearch/search/sort/FieldSortBuilder.java?pr=/elastic/elasticsearch/pull/26395

martijnvg commented:
retest this please

martijnvg commented:
@jpountz Your idea worked :) 

I'll squash and merge once the build is green.

mattweber commented:
Wow lots going on.  Thanks @martijnvg @jpountz @jasontedor!  Let me know if you need me to do anything.

martijnvg commented:
backported to 6.x branch via: 2425da1ccd65846c0fb2721cfe6c2e21d2f77e34 and 42c774583a52fc3856615747f14c3750db43d560

Changed the version check in master via ea3fa768f95a0645684e153d41563bcd9117fd1b
